polyfill/test262-harness.mjs

```diff
@@ -0,0 +1,246 @@
+import process from 'node:process';
+import ProgressBar from 'progress';
+import util from 'node:util';
+import vm from 'node:vm';
+import yaml from 'js-yaml';
+
+// = Temporal test262 runner =
+
+// This is a lightweight runner for Temporal's test262 tests. It exists in order
+// to be fast, and it is specific to Temporal. It doesn't provide all of the
+// facilities that the full test262 suite would expect. It achieves the "fast"
+// goal by parsing the Temporal polyfill once into a vm.Script, instead of
+// reading and parsing the whole file once for each test like the prelude option
+// of test262-harness would.
+
+// Without any arguments, all of the Temporal tests will be run. The script
+// takes one argument: a glob pattern relative to Temporal/ that specifies a
+// subset of tests to be run. For example:
+//   "PlainDateTime/**"
+//   "*/prototype/with/*.js"
+//   "**/basic.js"
+
+// For code coverage, run it with the environment variable NODE_V8_COVERAGE set
+// to a path relative to the cwd, where coverage metrics will be output. These
+// can be processed with the c8 tool.
+
+// Note, as of Node 18.7 there is a memory leak that makes it impossible to run
+// the entire suite with NODE_V8_COVERAGE, so you should run it in chunks.
+
+// Time the whole thing from start to finish
+const start = process.hrtime.bigint();
+
+// === Utilities and constants ===
+
+function print(str) {
+  process.stdout.write(str + '\n');
+}
+
+// Fancy output only if stdout is a terminal
+color.enabled = process.stdout.isTTY;
+
+// Front matter consists of a YAML document in between /*--- and ---*/
+const frontmatterMatcher = /\/\*---\n(.*)---\*\//ms;
+
+const UTF8 = { encoding: 'utf-8' };
+
+// === Preparation ===
+
+// Prepare Temporal polyfill. This vm.Script gets executed once for each test,
+// in a fresh VM context.
+
+const polyfillCode = fs.readFileSync('script.js', UTF8);
+const polyfill = new vm.Script(polyfillCode, { filename: path.resolve('script.js') });
+
+// Read the expected failures file and put the paths into a Set
+
+const expectedFailures = new Set(
+  fs
+    .readFileSync('test/expected-failures.txt', UTF8)
+    .split(/\r?\n/g)
+    .filter((line) => line && line[0] !== '#')
+);
+
+// This function reads in a test262 harness helper file, specified in 'includes'
+// in the frontmatter, and caches the resulting vm.Script so it can be used in
+// future tests that also include it.
+
+const helpersCache = new Map();
+function getHelperScript(includeName) {
+  if (helpersCache.has(includeName)) return helpersCache.get(includeName);
+
+  const includeFile = `test262/harness/${includeName}`;
+  const includeCode = fs.readFileSync(includeFile, UTF8);
+  const include = new vm.Script(includeCode);
+
+  helpersCache.set(includeName, include);
+  return include;
+}
+
+const testGlob = process.argv[2] ?? '**/*.js';
+
+const testFiles = (function () {
+  try {
+    return globSync(`test262/test/*/Temporal/${testGlob}`, { filesOnly: true });
```

Comment:
Should this ensure that each match looks like a test file rather than e.g. a README?
```suggestion
    return globSync(`test262/test/*/Temporal/${testGlob}`, { filesOnly: true }).filter(name => name.endsWith('js'));
```